File: ContrastControl.py
from cv2 import *
import numpy as np


def controller(sigmoid):
    processed = np.multiply(img, 1 / (np.exp(-np.multiply(np.multiply(img - 127.5, 1 / 255), 0.1 * sigmoid)) + 1)) + 0.5
    processed = np.clip(processed, 0, 255).astype(np.uint8)
    imshow('exp1-1_processed', processed)


img = imread('/home/liyang/图片/einstein.jpeg')
print(type(img), img.shape, img.dtype)

namedWindow('exp1-1_original', WINDOW_AUTOSIZE)
moveWindow('exp1-1_original', 200, 200)
imshow('exp1-1_original', img)

namedWindow('exp1-1_processed', WINDOW_AUTOSIZE)
moveWindow('exp1-1_processed', 800, 200)
imshow('exp1-1_processed', img)

createTrackbar('contrast control', 'exp1-1_processed', 0, 100, controller)

while True:
    k = waitKey(0)
    if k == 27:
        break

destroyAllWindows()

File: exp1/ContrastControl.cpp
#include <bits/stdc++.h>
#include <opencv2/opencv.hpp>
#include <cmath>

using namespace std;
using namespace cv;

int SigmoidControl;
Mat original, processed;

void ContrastControl(int, void *) {
    for (int y = 0; y < original.rows; y++)
        for (int x = 0; x < original.cols; x++)
            for (int c = 0; c < 3; c++) {
                double u = ((original.at<Vec3b>(y, x)[c] - 127.0) / 255.00) * SigmoidControl * 0.1;
                processed.at<Vec3b>(y, x)[c] = saturate_cast<uchar>(
                        original.at<Vec3b>(y, x)[c] * (1.00 / (1.00 + exp(-u))) + 0.4);
            }
    imshow("exp1-1_original", original);
    imshow("exp1-1_processed", processed);
}

int main(int argc, char **argv) {
    ios::sync_with_stdio(false);

    original = imread("/home/liyang/图片/einstein.jpeg");
    if (original.empty()) return -1;
    processed = Mat::zeros(original.size(), original.type());

    SigmoidControl = 100;

    namedWindow("exp1-1_original", WINDOW_AUTOSIZE);
    moveWindow("exp1-1_original", 200, 200);
    namedWindow("exp1-1_processed", WINDOW_AUTOSIZE);
    moveWindow("exp1-1_processed", 800, 200);
    createTrackbar("Contrast Controller", "exp1-1_processed", &SigmoidControl, 100, ContrastControl);
    ContrastControl(0, nullptr);

    waitKey(0);

    return 0;
}